tests/MongoDB.Driver.Tests/Specifications/change-streams/ChangeStreamTestRunner.cs

```diff
@@ -301,6 +302,20 @@ private List<ChangeStreamDocument<BsonDocument>> ReadChangeStreamDocuments(IAsyn
             }
 
             return result;
+
+            IEnumerable<IEnumerable<ChangeStreamDocument<BsonDocument>>> ReliableMoveNext()
+            {
+                var stopwatch = Stopwatch.StartNew();
+                while (async ? cursor.MoveNextAsync().GetAwaiter().GetResult() : cursor.MoveNext())
+                {
+                    if (cursor.Current.Any() || stopwatch.Elapsed > TimeSpan.FromSeconds(10))
+                    {
+                        yield return cursor.Current;
+                    }
+
+                    Thread.Sleep(TimeSpan.FromMilliseconds(300)); // delay between attempts
+                }
+            }
```

Comment:
I think this method can be refactored more simply as:

    private List<ChangeStreamDocument<BsonDocument>> ReadChangeStreamDocuments(IAsyncCursor<ChangeStreamDocument<BsonDocument>> cursor, BsonDocument test, bool async)
    {
        var result = new List<ChangeStreamDocument<BsonDocument>>();
        var resultDocument = test["result"].AsBsonDocument;
        var successNode = resultDocument["success"]?.AsBsonArray;

        var stopwatch = Stopwatch.StartNew();
        while (async ? cursor.MoveNextAsync().GetAwaiter().GetResult() : cursor.MoveNext())
        {
            result.AddRange(cursor.Current);

            if (successNode != null && result.Count >= successNode.Count)
            {
                break;
            }

            if (stopwatch.Elapsed > TimeSpan.FromSeconds(10))
            {
                break;
            }
        }

        return result;
    }

Reply:
Main points:

* the main problem was that expectedNumberOfDocuments should not be used unless "success" is present
* there is no need for a 300ms delay between calls to MoveNext
* the previous code had the potential to be an infinite loop with the stopwatch now prevents

Reply:
done